### ISSUE TITLE
Hufnagel Examples Needed

### ISSUE THREAD

**Issue by kyrieb-ekat** [DDMAL/Neon2#1353]
Let's provide links in this thread to provide a proto-ground truth for Zih Syuan what Hufnagel looks like. Since this is for developing a font for Neon, we're needing to think "classic" or idiomatic look of later Hufnagel script. The skinnier/more cursive versions are not what we need right now, though perhaps in the future that will change ([example](https://www.e-codices.unifr.ch/en/csg/0392/15/0/)). 

An example of what I mean by classic/idiomatic is this fellow ([e-codices link](https://www.e-codices.unifr.ch/en/list/one/csg/0438)). Examples can come from Cantus Database or general perusal; this source is in codices listed as having "german plainsong notation". 

<img width="590" height="807" alt="Image" src="https://github.com/user-attachments/assets/e1c0cfc7-5ca5-40bd-af7a-34963136987c" />
<img width="512" height="698" alt="Image" src="https://github.com/user-attachments/assets/51e063f3-c9f3-478c-ae17-2607c2c66356" />

**Comment by JoyfulGen** [comment]
I submit these for approval:

CDN-Mrb MS Medieval 0025: https://fragmentarium.ms/view/page/F-mtrk/10143/74902

NZ-Wt MSR-03: https://ndhadeliver.natlib.govt.nz/delivery/DeliveryManagerServlet?dps_pid=IE2339587&dps_custom_att_1=emu

Cantus MA Impr. 1537: https://daten.digitale-sammlungen.de/~db/0009/bsb00090351/images/index.html?seite=8&fip=193.174.98.30

The above three I found using my CantusDB spreadsheet. I also found this printed example while wandering the internets: https://www.cepuckett.com/inventory/index.php?no_cache=20260406152205&main_page=product_info&products_id=3061

**Comment by annamorphism** [comment]
https://public-content.library.mcgill.ca/digitization/rbsc_ms-medieval-234.pdf McGill's big Hufnagel manuscript could be good too...or https://diglib.hab.de/mss/29-helmst/start.htm. There's something kind of fun about saying we developed the font based on a manuscript at McGill...

Here's another print fragment, too: https://fragmentarium.ms/view/page/F-2ryf/5114/48789 . (The little lines of division aren't necessary for us right now). I wouldn't say the print pages are quite as typical, but the nice thing about the print examples is that their publishers were often trying to minimize the total number of required shapes to cast out of metal, which is similar to our situation (we'd prefer not to have too many shape-buttons).

**Comment by annamorphism** [comment]
@JoyfulGen pretty spreadsheet request: can you create some crops of each neume to get a representative glyph for each type?

**Comment by JoyfulGen** [comment]
@annamorphism is this what you had in mind? https://docs.google.com/spreadsheets/d/11Hfw4gD4au10WkWjswVRbE1t02NCWA3mBemxQ8zm7s0/edit?usp=sharing

Let me know if you'd like more neumes! If there are any manuscripts in there that you particularly like, I can add the full neume complement for those.

**Comment by annamorphism** [comment]
> [@annamorphism](https://github.com/annamorphism) is this what you had in mind? https://docs.google.com/spreadsheets/d/11Hfw4gD4au10WkWjswVRbE1t02NCWA3mBemxQ8zm7s0/edit?usp=sharing
> 
> Let me know if you'd like more neumes! If there are any manuscripts in there that you particularly like, I can add the full neume complement for those.

this is great! (i hadn't really noticed that one sample uses some sort of quarter rest for the f clef; how fun!). It pleases me that Cantus MA Impr. 1537 and St Gall 438 basically agree on the shapes, which makes me think those are the ones we should pursue in more depth. (None of them look like they will break into little constituent Neon-friendly neume components more easily than any other...)

**Comment by JoyfulGen** [comment]
Ok, I've added some neumes to the St Gall 438 and Cantus MA lines! I have some follow-up questions:

@zih-syuan, in the case of neumes that can have different intervals, can you infer bigger intervals from one example, or is it best to have a separate example of each interval? For example, the podatus (one note, and then another note higher than the first) can cover an interval of a second (podatus2), like this:

<img width="77" height="95" alt="Image" src="https://github.com/user-attachments/assets/aa66f96e-3a95-4697-8442-924641ad825e" />

Or an interval of a third (podatus3), like this:

<img width="76" height="127" alt="Image" src="https://github.com/user-attachments/assets/6440e2c8-6c22-45bf-8a1f-4f650665a5d9" />

Hypothetically, a podatus could also cover an interval of a fourth, or even a fifth. Is it possible for you to "stretch" the podatus3 to cover bigger intervals, or is it best if I find examples of a podatus4 and a podatus5? 

@annamorphism, @kyrieb-ekat, @zih-syuan, how do you feel about compound neumes that can be assembled from components we already have? I have several examples to submit:

<img width="113" height="105" alt="Image" src="https://github.com/user-attachments/assets/3c8c7f09-995b-44b7-9985-0812583a9f1a" /> 

The climacus - could be a virga and some puncta;

<img width="100" height="85" alt="Image" src="https://github.com/user-attachments/assets/8cd44570-48b0-4984-bb43-9fad37d82d34" />

The podatus subpunctis - could be a podatus and some puncta;

<img width="71" height="77" alt="Image" src="https://github.com/user-attachments/assets/504a0338-d6a7-4765-a334-a9f7e0c54832" />

The scandicus - could be some puncta and a virga;

<img width="127" height="101" alt="Image" src="https://github.com/user-attachments/assets/7c3229e0-56d0-4c67-85b5-030777240cdb" />

The torculus resupinus - could a torculus and a virga.

This is already the way that Neon builds most compound neumes in square notation (a podatus is two puncta, a clivis is a reversed virga and a punctum, etc.), so I think it could be great for hufnagel too. Especially since hufnagel already has a lot more symbols, what with the podatus and clivis being distinct shapes, building the bigger neumes from smaller components would avoid having a symbol library that's 2 pages long!

And finally, @annamorphism, what is this? Is it a liquescent? If so, does it go up or down?
<img width="47" height="52" alt="Image" src="https://github.com/user-attachments/assets/1a4f5d08-6561-40f9-ad7b-22528278eb39" />


And what is this? Is it a torculus with a liquescent attached to it?
<img width="148" height="95" alt="Image" src="https://github.com/user-attachments/assets/bc563671-6e97-446f-9af7-9038fdca0f1a" />

**Comment by kyrieb-ekat** [comment]
1) Gen this is amazing, very beautiful, great work
2) re: compound neumes; I vote yes? They're part and parcel of the neume lexicon, so I feel like it'd be weirder _not_ to have a climacis, porrectus, or scandicus and so on. 

> And what is this? Is it a torculus with a liquescent attached to it?
<img width="296" height="190" alt="Image" src="https://github.com/user-attachments/assets/70065642-199b-4000-93df-68f09ec07846" />

kinda winder is this fellow is just a torculus with a rising plica, or just a swanky lift? torculus epiphonus? does tha exist? 🤔 

<img width="94" height="104" alt="Image" src="https://github.com/user-attachments/assets/f8c718da-b583-4e1e-88ee-8680f3cc939f" />

I know this one!! maybe!! A pes epiphonus? 👀 

<img width="352" height="400" alt="Image" src="https://github.com/user-attachments/assets/74dff0c9-5da6-4c64-b443-d3b68d3e517d" />

@annamorphism , thoughts?

**Comment by annamorphism** [comment]
> 1. Gen this is amazing, very beautiful, great work
> 
>     2. re: compound neumes; I vote yes? They're part and parcel of the neume lexicon, so I feel like it'd be weirder _not_ to have a climacis, porrectus, or scandicus and so on.

I mean, kinda, but in the square note MSS we mostly ignored the neume shape labels in Neon and just thought of them as being composed of neume components (square, diamond, oblique, with or without stem)...But maybe going forward it would be useful to make more use of this vocabulary?
Also, as long as we're chopping things up, the torculus is basically a clivis with a punctum on the front.


> > And what is this? Is it a torculus with a liquescent attached to it?



> kinda winder is this fellow is just a torculus with a rising plica, or just a swanky lift? torculus epiphonus? does tha exist? 🤔

> 

does it need a name? torculus praepunctus liquescens?



> <img alt="Image" width="94" height="104" src="https://private-user-images.githubusercontent.com/74431986/588549359-f8c718da-b583-4e1e-88ee-8680f3cc939f.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NzgxMDQ5MzMsIm5iZiI6MTc3ODEwNDYzMywicGF0aCI6Ii83NDQzMTk4Ni81ODg1NDkzNTktZjhjNzE4ZGEtYjU4My00ZTFlLTg4ZWUtODY4MGYzY2M5MzlmLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNjA1MDYlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjYwNTA2VDIxNTcxM1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTg4NGNkZTc4NjUyM2Y3MDc5N2M3ZDM2ODcxNjU4YTZkYzlkZGQxOWY0NDY0OTUyMzY0ZWZiNDVhMTMzOTY0YjAmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JnJlc3BvbnNlLWNvbnRlbnQtdHlwZT1pbWFnZSUyRnBuZyJ9.zP3Mupw96ZrVdgIPWT113iZ5hh5uo_PlRbV_VcBWWH0">
I'd call this a liquescent going up (epiphonus)

**Comment by JoyfulGen** [comment]
To continue the compound neume conversation, @kyrieb-ekat I hear you that it's weird not to have certain neumes, but my concern is that the neume insert menu is going to get real big and confusing real fast. As Anna pointed out, the current square notation Neon allows the user to build neumes out of neume components. So if the user wants to input a podatus, they can input two puncta and group them. Once they're grouped, Neon _will_ recognize the neume as a podatus and identify it as such:

https://github.com/user-attachments/assets/6f68b4af-469d-4cae-8ca6-a160c0e7e300

With Hufnagel, we already can't build podati and clives from neume components, because they're their own unique shape. On top of that, the two notes of each shape are attached; unlike with the two puncta of a square notation podatus, the user can't individually move either pitch of a Hufnagel podatus. Because of that, I assume we'll need one shape per interval in the insert menu: a podatus2, a podatus3, a clivis2, a clivis3, etc. The advantage of having compound neumes as individual components is that it keeps the insert menu fairly straightforward and streamlined, and that the user can easily change the interval between different notes of the neume without having to insert a new neume. We can keep the same system wherein as soon as the components are grouped, Neon identifies them as one neume and gives that neume its correct name. What do you think?

**Comment by kyrieb-ekat** [comment]
I am convinced by this! I suppose my only remaining concern are combined neumes that are visually distinct from the otherwise combined form—where two tinier components are combined and do the little sailor moon transformation in Neon into a porrectus, we still need the porrectus icon to show up, yes? I may be misunderstanding the original question

**Comment by annamorphism** [comment]
>  the little sailor moon transformation in Neon

feature request!

>we still need the porrectus icon to show up, yes?

well, we have some things under "groupings" but the name in the "info" shown above is created from the components following an up-down-up pattern or whatever. If I group a clivis with a lower punctum, it creates the name climacus; if I pull the final one up so it's higher, it's now named porrectus, without my ever interacting with a porrectus icon.

**Comment by zih-syuan** [comment]
Regarding the handling of compound neumes, I would suggest adopting a separation between semantic representation and visual rendering.

My current understanding is that the underlying data should remain atomic, although I am still learning the details of neume representation. Even if the rendered result appears as a visually connected compound glyph, the user would still manipulate individual neume components, such as `punctum` or `virga`. When the interval changes, the system could dynamically replace the displayed SVG with the corresponding graphical variant.

The key requirement for this approach is preparing a sufficient set of SVG assets. I would need someone with more knowledge of the final implementation requirements to decide which neume types and interval combinations should be supported, for example `podatus2`, `podatus3`, and so on, so that I can prepare the corresponding graphical files.

This might also help keep the menu simpler.

**Comment by kyrieb-ekat** [comment]
I have moved this issue to "blocked" on the board, as it's somewhat in suspension until other methodological items are resolved, pending further discussions.

**Comment by Yueqiao12Zhang** [comment]
The new IC built-in training dataset is also looking for a manually labeled Hufnagel dataset to test performance.

**Comment by kyrieb-ekat** [comment]
I've done a quick annotation run at the hufnagel sample "Antiphonal_12v", attached below. As a general note to folks, I like to use the Virtual Geometry Group (VGG)'s "V(GG) Image Annotator" (VIA) for rapid detail annotation. A little bit initially confusing but it's easy to use quickly. https://annotate.officialstatistics.org/

[rapid_hufnagel_annotation.csv](https://github.com/user-attachments/files/28067562/rapid_hufnagel_annotation.csv)

<img width="795" height="1145" alt="Image" src="https://github.com/user-attachments/assets/6beaa644-b355-45e5-adb1-2dc196ddb2a7" />

**Comment by Yueqiao12Zhang** [comment]
> I've done a quick annotation run at the hufnagel sample "Antiphonal_12v", attached below. As a general note to folks, I like to use the Virtual Geometry Group (VGG)'s "V(GG) Image Annotator" (VIA) for rapid detail annotation. A little bit initially confusing but it's easy to use quickly. https://annotate.officialstatistics.org/
> 
> [rapid_hufnagel_annotation.csv](https://github.com/user-attachments/files/28067562/rapid_hufnagel_annotation.csv)
> 
> <img alt="Image" width="795" height="1145" src="https://private-user-images.githubusercontent.com/74431986/595602581-6beaa644-b355-45e5-adb1-2dc196ddb2a7.jpg?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NzkzNzE3MzcsIm5iZiI6MTc3OTM3MTQzNywicGF0aCI6Ii83NDQzMTk4Ni81OTU2MDI1ODEtNmJlYWE2NDQtYjM1NS00NWU1LWFkYjEtMmRjMTk2ZGRiMmE3LmpwZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNjA1MjElMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjYwNTIxVDEzNTAzN1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWQwMTkzYTIyY2Q3NGE0OGM2OWEyYmQ5YTkzOGNlMzhjY2FmYzdhODY3MzY3OGYxZmE1OWI3YmNhZTAxMjUwMzkmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JnJlc3BvbnNlLWNvbnRlbnQtdHlwZT1pbWFnZSUyRmpwZWcifQ.u7mCbO1CB1MGkh1QP70L7Pwr4SD3snUwDQ9mELpTU-0">

Thank you very much!

**Comment by JoyfulGen** [comment]
@kyrieb-ekat I feel that this issue can be closed now; do you agree?

**Comment by kyrieb-ekat** [comment]
Yes!
